Add support for parameter ranges and disabling readonly parameters.

malban opened the pull request:
In ROS2 it's possible to define parameter constraints min, max, and step size as well as if a parameter is read only or not.

This PR updates the tool to query for the parameter descriptions using the "describe_parameters" service (https://docs.ros2.org/foxy/api/rcl_interfaces/srv/DescribeParameters.html) and use the description to constrain the input widgets.

If a range is defined for either int or float parameters, a slider input with the range is used instead of a drag.  If a non-zero step size is defined, then the slider will snap to the nearest valid value, which is in the form: min + N x step.

It's important to honor the range and step constraints because any requests to set values outside the constraints are rejected by ROS.

Any read-only parameters are disabled so it's clear to the user that they can't be modified.

Before:
![before](https://github.com/user-attachments/assets/53d5437d-a4ed-48ea-9b1f-c8becf36a13a)

After:
![after2](https://github.com/user-attachments/assets/4b423fb6-3e44-40c2-8dbe-a7191f8c0d84)

ottojo commented on src/parameter_window.cpp:
```diff
@@ -108,6 +108,17 @@ void renderParameterWindow(const char *windowName, const std::string &curSelecte
     ImGui::End();
 }
 
+std::string getFormatStringFromStep(double step, int max_digits = 10) {
+    const double epsilon = 1e-12;  // tolerance for rounding error
+    int digits = 0;
+    double scaled = step;
+    while (digits < max_digits && std::abs(scaled - std::round(scaled)) > epsilon) {
+        scaled *= 10.0;
+        ++digits;
+    }
+    return "%." + std::to_string(digits) + "f";
+}
```
i'm not sure if building a format string at runtime and immediately parsing it again is the best solution here, maybe we could use fmtlib (is it already used in this project?)? it should allow dynamic precision easily (see the "dynamic precision example here: https://fmt.dev/11.1/syntax/#format-examples)

malban replied:
setting the format string of the slider to have the same precision as the step size is kind of a hack to account for the fact that the ImGui slider doesn't have a step setting.  It's not necessary, but it looks a lot nicer than having something like an arbitrary 6 decimal points of precision on the slider when it's just going to get snapped to the step size anyway.

I think the ImGui slider call is expecting an actual format string, so it needs to be generated regardless if it's dynamic. But if fmtlib is available it could be constructed like this, if that's preferred:

fmt::format("%.{:d}f", digits);

ottojo replied:
> I think the ImGui slider call is expecting an actual format string

i didn't realize imgui does the format string processing internally, i thought we just provide it with a formatted string. fmt doesnt seem to be used here yet, so let's leave it like it is to keep this PR small

ottojo commented on include/requests.hpp:
```diff
@@ -28,7 +32,18 @@ struct Request {
+              parameterDescriptors(parameterDescriptors_) {};
+
+    std::vector<std::string> parameterNames;
+    std::vector<rcl_interfaces::msg::ParameterDescriptor> parameterDescriptors;
```
why does the value request need the descriptors?

malban replied:
This is for convenience so that the descriptors can be passed on when the values are received:

[service_wrapper.cpp:L203](https://github.com/teamspatzenhirn/rig_reconfigure/pull/42/files#diff-237bafec7ef58175842b31b7c6716d0e1364dcce3e4bf36df986c82abf0f6e76R203)

and then here

[service_wrapper.cpp:L300](https://github.com/teamspatzenhirn/rig_reconfigure/pull/42/files#diff-237bafec7ef58175842b31b7c6716d0e1364dcce3e4bf36df986c82abf0f6e76R300)

Open to other ways of doing this.

ottojo replied:
no problem if it makes sense, just seemed nonintuitive when looking at it

ottojo reviewed (commented):
just some notes, looks good to me, needs a review from @authaldo

ottojo commented:
just another note/question: how does this behave if one of the bounds is something like `std::numeric_limits<double>::max()` ?
i think that happens for example when using the generate_parameter_library and specifying a condition like "must be >0" or similar

https://github.com/PickNikRobotics/generate_parameter_library/blob/13a1043416b62393d284f067c0c52d4db547a175/generate_parameter_library_py/generate_parameter_library_py/jinja_templates/cpp/declare_parameter#L15-L36

malban commented:
> just another note/question: how does this behave if one of the bounds is something like `std::numeric_limits<double>::max()` ? i think that happens for example when using the generate_parameter_library and specifying a condition like "must be >0" or similar
> 
> https://github.com/PickNikRobotics/generate_parameter_library/blob/13a1043416b62393d284f067c0c52d4db547a175/generate_parameter_library_py/generate_parameter_library_py/jinja_templates/cpp/declare_parameter#L15-L36

I'm not sure.  It probably makes sense to use a drag for ranges that are effectively unbounded on one or both sides.   Even with a drag, a non-zero step size could be enforced.  And if one side of the range is bounded, it could be clamped to that.

I'll take a look at this.

ottojo commented on include/parameter_tree.hpp:
```diff
@@ -12,6 +12,9 @@
 #include <memory>
 #include <optional>
 #include <utility>
+
+#include <rcl_interfaces/msg/parameter_descriptor.hpp>
+
```
no objections, i have no interest in ros1, this is 100% ros (2) specific anyways

ottojo commented:
> > just another note/question: how does this behave if one of the bounds is something like `std::numeric_limits<double>::max()` ? i think that happens for example when using the generate_parameter_library and specifying a condition like "must be >0" or similar
> > https://github.com/PickNikRobotics/generate_parameter_library/blob/13a1043416b62393d284f067c0c52d4db547a175/generate_parameter_library_py/generate_parameter_library_py/jinja_templates/cpp/declare_parameter#L15-L36
> 
> I'm not sure. It probably makes sense to use a drag for ranges that are effectively unbounded on one or both sides. Even with a drag, a non-zero step size could be enforced. And if one side of the range is bounded, it could be clamped to that.
> 
> I'll take a look at this.

i think even just adding the "effectively bounded" to the check deciding between drag and slider would be enough for now, as that would improve usability for `bounded_left && bounded_right` (and optionally with step size), and keep all the other cases as they are currently

but if you want to look at handling cases where one or both sides are effectively unbounded as well, please go ahead :smile:

malban commented:
> > > just another note/question: how does this behave if one of the bounds is something like `std::numeric_limits<double>::max()` ? i think that happens for example when using the generate_parameter_library and specifying a condition like "must be >0" or similar
> > > https://github.com/PickNikRobotics/generate_parameter_library/blob/13a1043416b62393d284f067c0c52d4db547a175/generate_parameter_library_py/generate_parameter_library_py/jinja_templates/cpp/declare_parameter#L15-L36
> > 
> > 
> > I'm not sure. It probably makes sense to use a drag for ranges that are effectively unbounded on one or both sides. Even with a drag, a non-zero step size could be enforced. And if one side of the range is bounded, it could be clamped to that.
> > I'll take a look at this.
> 
> i think even just adding the "effectively bounded" to the check deciding between drag and slider would be enough for now, as that would improve usability for `bounded_left && bounded_right` (and optionally with step size), and keep all the other cases as they are currently
> 
> but if you want to look at handling cases where one or both sides are effectively unbounded as well, please go ahead 😄

I updated the code to handle all the cases I could think of:
- Added a 'hasBoundedRange' function to check if it should be a slider or drag.
- Both the drag and slider will comply with the bounds and step if set.
- I tried to mirror how rclcpp validates values: https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L197C1-L197C33
- There is at least one poorly defined case, which is if the min value is std::numeric_limits<double>::lowest() and there is a non-zero step.   rclcpp is unable to validate valid values in case because of overflowing when subtracting the min value.   This is more of an issue with how the ranges are allowed to be defined, though.

authaldo reviewed (approved):
Only two minor requests:
 - the utility functions would in my view fit perfectly into the utility header / source files
 - the variable `curWindowWidth` in line 52 of `src/parameter_window.cpp` seems to be unused now (we potentially missed that in the last PR) and could be removed
 
 Only downside I found is that with activated limits there is no possibility to enter a value directly. However, this should only be critical for large ranges and thus hopefully only seldomly an issue. I am not sure whether ImGui even has support for entering values in dragScalars.
 
 Thanks again for the nice contribution!

malban commented:
> Only downside I found is that with activated limits there is no possibility to enter a value directly. However, this should only be critical for large ranges and thus hopefully only seldomly an issue. I am not sure whether ImGui even has support for entering values in dragScalars.

I just looked into this and it turns out that you can enter values directly into the sliders and drags with ctrl+click.  For the drags it appears double clicking will also work.

authaldo commented:
> > Only downside I found is that with activated limits there is no possibility to enter a value directly. However, this should only be critical for large ranges and thus hopefully only seldomly an issue. I am not sure whether ImGui even has support for entering values in dragScalars.
> 
> I just looked into this and it turns out that you can enter values directly into the sliders and drags with ctrl+click. For the drags it appears double clicking will also work.

Nice! I honestly never checked anything else than double clicking since that was the behaviour I've been used to. 

Other minor finding:
Could it be that directly inserting the value bypasses the ImGui limits? 
We already have kind of a similar issue with #30, where we keep the entered value although the node rejected the update (currently we do not keep track of the old value to update the visualization in that case). 
However, knowing the limits beforehand simplifies the handling of these cases. 

The two flags suggested [here](https://github.com/ocornut/imgui/issues/1829#issuecomment-698913609) could potentially fix the issue.

malban commented:
Updated

ottojo commented:
I finally did some testing, looks really good! i discovered an upstream bug with the generate_parameter_library which i reported (https://github.com/PickNikRobotics/generate_parameter_library/issues/256) and some weird issue with half-open integer ranges, but i suggest merging this and i will look at fixing that one in a followup.

@authaldo any objections to merging?
